src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRun/ModuleTokenResolver.cs

```diff
@@ -139,11 +139,14 @@ public void AddModuleTokenForField(FieldDesc field, ModuleToken token)
                 return;
             }
 
-            _fieldToRefTokens[field] = token;
+            TypeDesc owningCanonType = field.OwningType.ConvertToCanonForm(CanonicalFormKind.Specific);
+            FieldDesc canonField = owningCanonType.GetField(field.Name);
```

Comment:
```suggestion
            FieldDesc canonField = field;
            if (owningCanonType != field.OwningType)
                canonField = _typeSystemContext.GetFieldForInstantiatedType(field.GetTypicalFieldDefinition(), (InstantiatedType)owningCanonType);
```

This is faster than doing the name lookup and it will stay correct once we add support for fields that only differ by signature (it's not allowed in C#, but ECMA-335 allows that).

Reply:
Thanks, I like the suggestion; I tried to apply it but I hit an interesting issue. There's a second place in SignatureBuilder using a similar construct to convert a field to canonical so I naturally modified it as you suggested. That however made the test ldtokenmember start crashing again - the problem is that, when the field is runtime-determined, in my case it's something like

<pre>
T_System.__Canon GenericHolder`2<T_System.__Canon, U_System.__Canon>.StaticFieldT
</pre>

and the call to GetFieldOnInstantiatedType (for GenericHolder`2<System.__Canon, System.__Canon>) doesn't change the field type from T_System.__Canon to just System.__Canon so that it subsequently fails the reverse token lookup. Is this a deficiency of the method GetFieldForInstantiatedType or do we need additional magic to get rid of the runtime-determined field type?

Reply:
Can you push that change out so I can have a look?

Reply:
Pushed. Thanks Michal! I have summarized the issue and the test I'm seeing fail with the issue in the commit message.